### PR TITLE
Remove support for extensionless import 

### PR DESCRIPTION
**Pull request by ry**
#1384

@kevinkassimo @kitsonk Please review

### CODE REVIEW

File: src/deno_dir.rs [hunk 1]
```diff
@@ -182,33 +177,24 @@ impl DenoDir {
...
-        filename: filename.clone(), // TODO: no clone after NLL rfc
-        media_type: map_content_type(&p, maybe_content_type_str),
-        source_code,
-        maybe_output_code: None,
-        maybe_source_map: None,
-      }));
-    }
-    Ok(None) // cannot find locally
+    let p = Path::new(&filename);
+    let media_type_filename = [&filename, ".mime"].concat();
```
Comment (kevinkassimo) [inline src/deno_dir.rs]: This has a very interesting implication: if a user provides a `.mime` file containing typescript mime type, s/he can use the extensionless file as if it has the .ts extension...

File: src/deno_dir.rs [hunk 2]
```diff
@@ -222,19 +208,35 @@ impl DenoDir {
...
+      match self.fetch_local_source(&module_name, &filename) {
+        Ok(Some(output)) => {
+          debug!("found local source ");
+          return Ok(output);
+        }
+        Ok(None) => {
+          debug!("NOT found local source ");
+        }
+        Err(err) => {
+          debug!(" Got Error from fetch_local_source {}", err);
```
Comment (kevinkassimo) [inline src/deno_dir.rs]: nit: remove extra space

File: js/deno.ts [hunk 3]
```diff
@@ -55,6 +55,13 @@ export { run, RunOptions, Process, ProcessStatus } from "./process";
 export { inspect } from "./console";
 export const args: string[] = [];
 
+// TODO Don't expose Console nor stringifyArgs.
```
Comment (kitsonk) [inline js/deno.ts]: We may have to do a _test_ build of Deno, like we do for Rust to deal with some of these unit tests in future.

Comment (kitsonk) [reply js/deno.ts]: I would also mark all of these with `@internal` so they don't leak into the `lib.deno_runtime.d.ts` like the compiler below.

File: js/globals.ts [hunk 4]
```diff
@@ -55,8 +55,12 @@ window.setInterval = timers.setInterval;
 // being used, which it cannot statically determine within this module.
 window.Blob = blob.DenoBlob;
 export type Blob = blob.DenoBlob;
-window.File = file.DenoFile;
-export type File = file.DenoFile;
+
+// TODO(ry) Do not export a class implementing the DOM, export the DOM
```
Comment (kitsonk) [inline js/globals.ts]: You should be able to do something like this, following the pattern below:

https://github.com/denoland/deno/blob/e1faccc83e5e00a97bad082df21a499cda2d927c/js/globals.ts#L72-L73

You will need a `FileConstructor` interface in `domTypes` though.

File: src/deno_dir.rs (hunk 1 above)
Comment (ry) [reply src/deno_dir.rs]: Hmm do we even need the mime type files anymore ?

Comment (kitsonk) [reply src/deno_dir.rs]: But the compiler will have issues with it (e.g. why the shebang scripts don't work without extensions at the moment).

File: js/deno.ts (hunk 3 above)
Comment (ry) [reply js/deno.ts]: Unfortunately that doesn't work:
```
/Users/rld/src/deno/js/net_test.ts:4:10 - error TS2305: Module '"deno"' has no exported member 'deferred'.

4 import { deferred } from "deno";
           ~~~~~~~~
```
However, I have added JSDOC comments that note these will be removed.

File: src/deno_dir.rs (hunk 2 above)
Comment (ry) [reply src/deno_dir.rs]: Fixed

File: js/globals.ts (hunk 4 above)
Comment (ry) [reply js/globals.ts]: I will leave this for a follow up PR as it's out of scope WRT removing extension support. I've added a link to your comment in the source.

File: src/deno_dir.rs (hunk 1 above)
Comment (ry) [reply src/deno_dir.rs]: I'd like to clean up the mime file usage in the future. It seems we don't need most of them. Leaving as is for now.

**Comment by piscisaureus** [comment]
> Have fetch_local_source return None instead of NotFound on error

What motivates this?

File: js/unit_tests.ts [hunk 5]
```diff
@@ -2,16 +2,16 @@
...
 import "./buffer_test.ts";
 import "./chmod_test.ts";
 import "./compiler_test.ts";
 import "./console_test.ts";
 import "./copy_file_test.ts";
-import "./dir_test";
+import "./dir_test.ts";
 import "./fetch_test.ts";
-import "./file_test.ts";
+// import "./file_test.ts";
```
Comment (piscisaureus) [inline js/unit_tests.ts]: why commented out?

**Comment by ry** [comment]
> > Have fetch_local_source return None instead of NotFound on error
> What motivates this?

Because `fetch_local_source` returns a `DenoResult<Option<CodeFetchOutput>>` - which is already `Ok(None)` when not found in other conditions.

**Comment by ry** [comment]
There was also this out-of-band discussion 
```
[11:23, 12/23/2018] Bert Belder: this patch is full of seemingly unrelated changes
[11:23, 12/23/2018] Ryan: It's because there's concept leakage that happened
[11:23, 12/23/2018] Ryan: js/*_test.ts file are run inside of deno
[11:23, 12/23/2018] Ryan: while the rest are run through rollup
[11:24, 12/23/2018] Ryan: so there's two different resolution algorithms being used
[11:24, 12/23/2018] Ryan: when we weakened the resolution scheme some js/*_test.ts were able import internal deno modules directly - without going thru the runtime
```

**Pull request merged**